File: include/extras-ktexteditor/libcursor.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/ktexteditor-cursor.html
pub const ktexteditor__cursor = struct {
    /// New constructs a new KTextEditor::Cursor object.
    ///
    /// ``` other: QtC.KTextEditor__Cursor ```
    pub fn New(other: ?*anyopaque) QtC.KTextEditor__Cursor {
        return qtc.KTextEditor__Cursor_new(@ptrCast(other));
    }

    /// New2 constructs a new KTextEditor::Cursor object and invalidates the source KTextEditor::Cursor object.
    ///
    /// ``` other: QtC.KTextEditor__Cursor ```
    pub fn New2(other: ?*anyopaque) QtC.KTextEditor__Cursor {
        return qtc.KTextEditor__Cursor_new2(@ptrCast(other));
    }

    /// New3 constructs a new KTextEditor::Cursor object.
    ///
    ///
    pub fn New3() QtC.KTextEditor__Cursor {
        return qtc.KTextEditor__Cursor_new3();
    }

    /// New4 constructs a new KTextEditor::Cursor object.
    ///
    /// ``` line: i32, column: i32 ```
    pub fn New4(line: i32, column: i32) QtC.KTextEditor__Cursor {
        return qtc.KTextEditor__Cursor_new4(@intCast(line), @intCast(column));
    }

    /// New5 constructs a new KTextEditor::Cursor object.
    ///
    /// ``` param1: QtC.KTextEditor__Cursor ```
    pub fn New5(param1: ?*anyopaque) QtC.KTextEditor__Cursor {
        return qtc.KTextEditor__Cursor_new5(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    ///
    /// ``` self: QtC.KTextEditor__Cursor, other: QtC.KTextEditor__Cursor ```
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.KTextEditor__Cursor_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    ///
    /// ``` self: QtC.KTextEditor__Cursor, other: QtC.KTextEditor__Cursor ```
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.KTextEditor__Cursor_MoveAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#isValid)
    ///
    /// ``` self: QtC.KTextEditor__Cursor ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.KTextEditor__Cursor_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#invalid)
    ///
    ///
    pub fn Invalid() QtC.KTextEditor__Cursor {
        return qtc.KTextEditor__Cursor_Invalid();
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#start)
    ///
    ///
    pub fn Start() QtC.KTextEditor__Cursor {
        return qtc.KTextEditor__Cursor_Start();
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#toString)
    ///
    /// ``` self: QtC.KTextEditor__Cursor, allocator: std.mem.Allocator ```
    pub fn ToString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KTextEditor__Cursor_ToString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("ktexteditor::cursor.ToString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#setPosition)
    ///
    /// ``` self: QtC.KTextEditor__Cursor, position: QtC.KTextEditor__Cursor ```
    pub fn SetPosition(self: ?*anyopaque, position: QtC.KTextEditor__Cursor) void {
        qtc.KTextEditor__Cursor_SetPosition(@ptrCast(self), @ptrCast(position));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#setPosition)
    ///
    /// ``` self: QtC.KTextEditor__Cursor, line: i32, column: i32 ```
    pub fn SetPosition2(self: ?*anyopaque, line: i32, column: i32) void {
        qtc.KTextEditor__Cursor_SetPosition2(@ptrCast(self), @intCast(line), @intCast(column));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#line)
    ///
    /// ``` self: QtC.KTextEditor__Cursor ```
    pub fn Line(self: ?*anyopaque) i32 {
        return qtc.KTextEditor__Cursor_Line(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#setLine)
    ///
    /// ``` self: QtC.KTextEditor__Cursor, line: i32 ```
    pub fn SetLine(self: ?*anyopaque, line: i32) void {
        qtc.KTextEditor__Cursor_SetLine(@ptrCast(self), @intCast(line));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#column)
    ///
    /// ``` self: QtC.KTextEditor__Cursor ```
    pub fn Column(self: ?*anyopaque) i32 {
        return qtc.KTextEditor__Cursor_Column(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#setColumn)
    ///
    /// ``` self: QtC.KTextEditor__Cursor, column: i32 ```
    pub fn SetColumn(self: ?*anyopaque, column: i32) void {
        qtc.KTextEditor__Cursor_SetColumn(@ptrCast(self), @intCast(column));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#atStartOfLine)
    ///
    /// ``` self: QtC.KTextEditor__Cursor ```
    pub fn AtStartOfLine(self: ?*anyopaque) bool {
        return qtc.KTextEditor__Cursor_AtStartOfLine(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#atStartOfDocument)
    ///
    /// ``` self: QtC.KTextEditor__Cursor ```
    pub fn AtStartOfDocument(self: ?*anyopaque) bool {
        return qtc.KTextEditor__Cursor_AtStartOfDocument(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-cursor.html#position)
    ///
    /// ``` self: QtC.KTextEditor__Cursor, line: *i32, column: *i32 ```
    pub fn Position(self: ?*anyopaque, line: *i32, column: *i32) void {
        qtc.KTextEditor__Cursor_Position(@ptrCast(self), @ptrCast(line), @ptrCast(column));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KTextEditor__Cursor ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KTextEditor__Cursor_Delete(@ptrCast(self));
    }
};

/// https://api.kde.org/ktexteditor.html
pub const ktexteditor = struct {};

File: src_original/extras-ktexteditor/libcursor.cpp
#include <KTextEditor/Cursor>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <cursor.h>
#include "libcursor.h"
#include "libcursor.hxx"

KTextEditor__Cursor* KTextEditor__Cursor_new(const KTextEditor__Cursor* other) {
    return new KTextEditor::Cursor(*other);
}

KTextEditor__Cursor* KTextEditor__Cursor_new2(KTextEditor__Cursor* other) {
    return new KTextEditor::Cursor(std::move(*other));
}

KTextEditor__Cursor* KTextEditor__Cursor_new3() {
    return new KTextEditor::Cursor();
}

KTextEditor__Cursor* KTextEditor__Cursor_new4(int line, int column) {
    return new KTextEditor::Cursor(static_cast<int>(line), static_cast<int>(column));
}

KTextEditor__Cursor* KTextEditor__Cursor_new5(const KTextEditor__Cursor* param1) {
    return new KTextEditor::Cursor(*param1);
}

void KTextEditor__Cursor_CopyAssign(KTextEditor__Cursor* self, KTextEditor__Cursor* other) {
    *self = *other;
}

void KTextEditor__Cursor_MoveAssign(KTextEditor__Cursor* self, KTextEditor__Cursor* other) {
    *self = std::move(*other);
}

bool KTextEditor__Cursor_IsValid(const KTextEditor__Cursor* self) {
    return self->isValid();
}

KTextEditor__Cursor* KTextEditor__Cursor_Invalid() {
    return new KTextEditor::Cursor(KTextEditor::Cursor::invalid());
}

KTextEditor__Cursor* KTextEditor__Cursor_Start() {
    return new KTextEditor::Cursor(KTextEditor::Cursor::start());
}

libqt_string KTextEditor__Cursor_ToString(const KTextEditor__Cursor* self) {
    QString _ret = self->toString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void KTextEditor__Cursor_SetPosition(KTextEditor__Cursor* self, KTextEditor__Cursor* position) {
    self->setPosition(*position);
}

void KTextEditor__Cursor_SetPosition2(KTextEditor__Cursor* self, int line, int column) {
    self->setPosition(static_cast<int>(line), static_cast<int>(column));
}

int KTextEditor__Cursor_Line(const KTextEditor__Cursor* self) {
    return self->line();
}

void KTextEditor__Cursor_SetLine(KTextEditor__Cursor* self, int line) {
    self->setLine(static_cast<int>(line));
}

int KTextEditor__Cursor_Column(const KTextEditor__Cursor* self) {
    return self->column();
}

void KTextEditor__Cursor_SetColumn(KTextEditor__Cursor* self, int column) {
    self->setColumn(static_cast<int>(column));
}

bool KTextEditor__Cursor_AtStartOfLine(const KTextEditor__Cursor* self) {
    return self->atStartOfLine();
}

bool KTextEditor__Cursor_AtStartOfDocument(const KTextEditor__Cursor* self) {
    return self->atStartOfDocument();
}

void KTextEditor__Cursor_Position(const KTextEditor__Cursor* self, int* line, int* column) {
    self->position(static_cast<int&>(*line), static_cast<int&>(*column));
}

void KTextEditor__Cursor_Delete(KTextEditor__Cursor* self) {
    delete self;
}
